### PR TITLE
Add files-mcp server sidecar

### LINKED ISSUE #1: Implement files-mcp MCP server
## User Request

Implement the `files-mcp` MCP server from scratch — a sidecar that gives agents access to platform files via a `read_file` tool.

## Specification

Per [files-mcp.md](https://github.com/agynio/architecture/blob/main/architecture/files-mcp.md) and [media.md](https://github.com/agynio/architecture/blob/main/architecture/media.md):

### Overview
- **Language:** Go
- **Transport:** Streamable HTTP
- **Role:** MCP server sidecar in agent pods — file access for agents
- **Single tool:** `read_file` with `file_id` input

### Tool: `read_file`

**Flow:**
1. Agent CLI sends `tools/call read_file {file_id}`
2. MCP server fetches file metadata from Files service (via Gateway) — `GetFileMetadata`
3. MCP server checks `size_bytes` against `MAX_FILE_SIZE`; if exceeded, returns error
4. MCP server calls `GetFileContent` on Files service (via Gateway) to stream content
5. MCP server builds MCP response based on MIME type
6. Returns tool result

**Response content types:**

| File MIME type | MCP content type | Format |
|----------------|-----------------|--------|
| `image/*` | `image` | `{ "type": "image", "data": "<base64>", "mimeType": "<content_type>" }` |
| `text/*`, `application/json`, `application/xml`, `application/yaml` | `text` | `{ "type": "text", "text": "<content>" }` |
| All other types | `resource` (binary) | `{ "type": "resource", "resource": { "uri": "agyn://file/<file_id>", "mimeType": "<content_type>", "blob": "<base64>" } }` |

**Error handling** (all errors as `isError: true` tool errors, NOT protocol-level):

| Condition | Error text |
|-----------|-----------|
| File not found | `file not found: <file_id>` |
| File exceeds max size | `file too large: <size_bytes> bytes exceeds limit of <max_bytes> bytes` |
| Download failure | `failed to download file: <details>` |
| Gateway/Files unavailable | `platform service unavailable` |

### Configuration

| Variable | Required | Description |
|----------|----------|-------------|
| `MCP_PORT` | yes | Port to listen on |
| `GATEWAY_ADDRESS` | yes | Gateway address (e.g., `gateway.ziti` or `https://gateway.agyn.dev`) |
| `AGYN_API_TOKEN` | no | API token for Gateway auth (dev/test fallback) |
| `MAX_FILE_SIZE` | no | Max file size in bytes (default: 20 MiB) |

### Authentication
- Primary: OpenZiti mTLS via Ziti sidecar (automatic in agent pods)
- Fallback: `AGYN_API_TOKEN` env var sent as `Authorization: Bearer`

### MCP Capabilities
```json
{ "capabilities": { "tools": {} } }
```
Static tool list — no `listChanged`, no resources, no prompts.

### Deliverables
- Full repo scaffolding: `cmd/`, `internal/`, `charts/`, `.github/workflows/`, `devspace.yaml`, `Dockerfile`, `README.md`
- Config, gRPC client to Gateway, MCP server with `read_file` tool
- Multi-arch Dockerfile (amd64 + arm64)
- Helm chart (service-base dependency)
- CI/CD workflows (ci.yml + release.yml)
- DevSpace config (dev + test-e2e pipelines)
- Unit tests + E2E tests

### Dependencies
- [agynio/api#86](https://github.com/agynio/api/pull/86) — `GetFileContent` proto
- [agynio/files#21](https://github.com/agynio/files/issues/21) — Files service `GetFileContent` implementation
- [agynio/gateway#134](https://github.com/agynio/gateway/issues/134) — Gateway `GetFileContent` proxy

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add files-mcp MCP server with `read_file` tool, gRPC gateway client, and content classification
- add Dockerfile, Helm chart, DevSpace config, and CI/release workflows
- add unit tests plus placeholder e2e coverage

## Testing
- buf generate buf.build/agynio/api --include-imports --path agynio/api/files/v1 --path agynio/api/gateway/v1/files.proto
- go vet ./...
- go test ./...

Closes #1

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- buf generate buf.build/agynio/api --include-imports --path agynio/api/files/v1 --path agynio/api/gateway/v1/files.proto
- go vet ./...
- go test ./...

Tests: passed 8, failed 0, skipped 0
Lint: go vet ./... (no issues)

File: internal/mcpserver/server.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+	"github.com/agynio/files-mcp/internal/contenttype"
+	"github.com/agynio/files-mcp/internal/filesclient"
+	"github.com/modelcontextprotocol/go-sdk/mcp"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const defaultMaxFileSize = int64(20 * 1024 * 1024)
+
+const (
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:21]: [major] Duplicate constant — `defaultMaxFileSize` is defined both here and in `internal/config/config.go`. This is a near-duplicate created by iterative AI edits. Define it in one place (config package is the natural owner as the boundary layer setting defaults) and reference it from here if needed, or simply remove the fallback in `New()` entirely (see next comment).

File: internal/mcpserver/server.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	maxSize := opts.MaxFileSize
+	if maxSize <= 0 {
+		maxSize = defaultMaxFileSize
+	}
+
+	server := mcp.NewServer(&mcp.Implementation{
+		Name:    "files-mcp",
+		Version: "0.1.0",
+	}, &mcp.ServerOptions{Capabilities: &mcp.ServerCapabilities{Tools: &mcp.ToolCapabilities{}}})
+
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:67]: [major] "Just-in-case" fallback in internal code. The config boundary (`config.FromEnv`) already validates that `MaxFileSize > 0` and applies the default. If `MaxFileSize <= 0` reaches `New()`, it is a programming error and should be surfaced loudly, not silently corrected.

Remove this fallback and panic (or return an error) if the invariant is violated:

```go
if opts.MaxFileSize <= 0 {
    panic("mcpserver.New: MaxFileSize must be positive")
}
```

This keeps the boundary/internal contract explicit. Also remove the duplicate `defaultMaxFileSize` from this file after this change.

File: internal/mcpserver/server.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	if err != nil {
+		return toolErrorResult(mapServiceError(err, input.FileID)), nil
+	}
+	if metadata == nil {
+		return toolErrorResult("failed to download file: missing metadata"), nil
+	}
+
+	sizeBytes := metadata.GetSizeBytes()
+	if sizeBytes > s.maxFileSize {
+		return toolErrorResult(fmt.Sprintf("file too large: %d bytes exceeds limit of %d bytes", sizeBytes, s.maxFileSize)), nil
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:106]: [major] Defensive nil check that contradicts the gRPC contract. If `GetFileMetadata` succeeds (no error), the response will contain a valid `FileInfo` — returning `nil` metadata without error is not part of the protobuf/gRPC contract. This is a "just-in-case" branch in internal logic that masks a real bug if it ever triggers.

Remove this check. If you want safety against a truly broken gateway, the correct approach is in the *boundary* (`filesclient.Client.GetFileMetadata`) — check there and return an explicit error if `resp.GetFile()` is nil, so internal code never sees it.

File: internal/mcpserver/server.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	content := buildContent(metadata.GetContentType(), input.FileID, data)
+	return &mcp.CallToolResult{Content: []mcp.Content{content}}, nil
+}
+
+func parseReadFileInput(req *mcp.CallToolRequest) (readFileInput, error) {
+	if req == nil || req.Params == nil || len(req.Params.Arguments) == 0 {
+		return readFileInput{}, errors.New("file_id is required")
+	}
+	var input readFileInput
+	if err := json.Unmarshal(req.Params.Arguments, &input); err != nil {
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:128]: [minor] The `req == nil || req.Params == nil` guards are boundary-level defensiveness inside internal handler code. The MCP SDK framework guarantees non-nil request/params when calling a `ToolHandler`. These checks add noise and suggest the contract is weaker than it actually is.

If you want to keep a guard, limit it to `len(req.Params.Arguments) == 0` (which checks for a legitimate "no arguments supplied" case).

File: internal/mcpserver/server.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+		return &mcp.ImageContent{Data: data, MIMEType: contentType}
+	case contenttype.KindText:
+		return &mcp.TextContent{Text: string(data)}
+	default:
+		return &mcp.EmbeddedResource{
+			Resource: &mcp.ResourceContents{
+				URI:      fmt.Sprintf("agyn://file/%s", fileID),
+				MIMEType: contentType,
+				Blob:     data,
+			},
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:151]: [major] "Just-in-case" nil check in internal code. gRPC streaming `Recv()` never returns `(nil, nil)` — it returns either a valid response or an error (including `io.EOF`). This `if resp == nil { continue }` branch silently swallows an impossible state.

Remove this guard. If a nil response somehow arrives, it should surface as a panic (via `resp.GetChunkData()`) rather than be silently skipped.

File: internal/mcpserver/server.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+		return readFileInput{}, errors.New("file_id is required")
+	}
+	return input, nil
+}
+
+func buildContent(contentType string, fileID string, data []byte) mcp.Content {
+	switch contenttype.Classify(contentType) {
+	case contenttype.KindImage:
+		return &mcp.ImageContent{Data: data, MIMEType: contentType}
+	case contenttype.KindText:
```
Comment (noa-lucent) [inline internal/mcpserver/server.go:143]: [minor] The `default` case in `buildContent` masks a potential invariant violation. `contenttype.Kind` is a closed enum with three values (`KindText`, `KindImage`, `KindResource`). Using `default` instead of `case contenttype.KindResource` means if a fourth kind is added, this function silently treats it as a resource instead of failing loudly.

Use an explicit `case contenttype.KindResource` and add a `default` that panics:

```go
case contenttype.KindResource:
    return &mcp.EmbeddedResource{...}
default:
    panic(fmt.Sprintf("buildContent: unhandled content kind %d", contenttype.Classify(contentType)))
```

File: internal/filesclient/client.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+func (BearerToken) RequireTransportSecurity() bool {
+	return false
+}
+
+func Dial(ctx context.Context, address string, token string) (*grpc.ClientConn, error) {
+	options := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
+	if token != "" {
+		options = append(options, grpc.WithPerRPCCredentials(BearerToken{Token: token}))
+	}
+	return grpc.DialContext(ctx, address, options...)
```
Comment (noa-lucent) [inline internal/filesclient/client.go]: [major] `grpc.DialContext` is deprecated since gRPC v1.67+ (you're on v1.79.2). Use `grpc.NewClient` instead:

```go
func Dial(address string, token string) (*grpc.ClientConn, error) {
    options := []grpc.DialOption{grpc.WithTransportCredentials(insecure.NewCredentials())}
    if token != "" {
        options = append(options, grpc.WithPerRPCCredentials(BearerToken{Token: token}))
    }
    return grpc.NewClient(address, options...)
}
```

Note: `grpc.NewClient` does not take a context — connection is established lazily. Remove the `ctx` parameter.

File: internal/filesclient/client.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+func New(conn *grpc.ClientConn) *Client {
+	return &Client{client: gatewayv1.NewFilesGatewayClient(conn)}
+}
+
+func (c *Client) GetFileMetadata(ctx context.Context, fileID string) (*filesv1.FileInfo, error) {
+	resp, err := c.client.GetFileMetadata(ctx, &filesv1.GetFileMetadataRequest{FileId: fileID})
+	if err != nil {
+		return nil, err
+	}
```
Comment (noa-lucent) [inline internal/filesclient/client.go:49]: [minor] `GetFileMetadata` returns `resp.GetFile()` which could be nil if the gateway sends an empty response (broken server). Since `filesclient` is the boundary layer to the external gRPC service, this is the right place to validate and reject that case:

```go
func (c *Client) GetFileMetadata(ctx context.Context, fileID string) (*filesv1.FileInfo, error) {
    resp, err := c.client.GetFileMetadata(ctx, &filesv1.GetFileMetadataRequest{FileId: fileID})
    if err != nil {
        return nil, err
    }
    file := resp.GetFile()
    if file == nil {
        return nil, fmt.Errorf("empty metadata response for file %s", fileID)
    }
    return file, nil
}
```

This keeps the nil-metadata validation in the boundary (where external data enters) and removes the need for the defensive nil check in `server.go` line 109–111.

File: internal/contenttype/contenttype.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+type Kind int
+
+const (
+	KindText Kind = iota
+	KindImage
+	KindResource
+)
+
+func Classify(contentType string) Kind {
+	mediaType := strings.TrimSpace(contentType)
```
Comment (noa-lucent) [inline internal/contenttype/contenttype.go:17]: [major] No unit tests for `Classify`. This function encodes critical business logic (the content-type-to-MCP-response-type mapping from the architecture spec) and has non-trivial parsing (media type normalization, case folding, mime.ParseMediaType). It must have its own test file covering:

- `image/*` variants (e.g., `image/png`, `image/jpeg`, `IMAGE/PNG` for case-insensitivity)
- `text/*` variants
- `application/json`, `application/xml`, `application/yaml`
- Binary types (`application/pdf`, `application/octet-stream`)
- Content types with parameters (e.g., `text/plain; charset=utf-8`)
- Empty / malformed input

File: internal/mcpserver/server_test.go [hunk 10]
```diff
@@ -0,0 +1,270 @@
+package mcpserver
```
Comment (noa-lucent) [inline internal/mcpserver/server_test.go:1]: [minor] Test coverage gaps. The existing tests are well-structured but miss a few important scenarios from the error handling spec:

1. **Download stream error** — `fakeFilesClient.streamErr` is set but never exercised. Add a test where `GetFileContent` succeeds but `Recv()` returns an error mid-stream, verifying the `"failed to download file: ..."` error path.
2. **`context.DeadlineExceeded`** — `isUnavailable` handles `context.DeadlineExceeded` but there's no test for it.
3. **`GetFileContent` connection error** — `contentErr` is set in the fake but never tested. Add a test where the content RPC itself fails (as opposed to the stream).

File: charts/files-mcp/templates/_helpers.tpl [hunk 11]
```diff
@@ -0,0 +1,29 @@
+{{- define "files-mcp.configureEnv" -}}
+{{- $env := list -}}
+
+{{- $mcpPort := int (default 8100 .Values.filesMcp.mcpPort) -}}
+{{- $env = append $env (dict "name" "MCP_PORT" "value" (printf "%d" $mcpPort)) -}}
+
+{{- $gatewayAddress := required "filesMcp.gatewayAddress is required" (trimAll " \n\t" (default "gateway.ziti" .Values.filesMcp.gatewayAddress)) -}}
```
Comment (noa-lucent) [inline charts/files-mcp/templates/_helpers.tpl]: [nit] The `required` guard is unreachable here — `default "gateway.ziti"` ensures the value is never empty, so `required` will never trigger. Either remove `required` (since there's a sensible default) or remove `default` (if you want to force explicit configuration). As written it's misleading.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Good implementation overall — the architecture spec is followed closely, the tool definition and error messages match, the MCP SDK usage is correct, and the infra (Dockerfile, Helm, CI, DevSpace) is solid.

Requesting changes for the following issues:

**Major:**
1. Duplicate `defaultMaxFileSize` constant across `config` and `mcpserver` packages.
2. "Just-in-case" fallback in `mcpserver.New()` — silent default instead of failing on invalid invariant.
3. Defensive nil metadata check in internal code (`server.go:109-111`) — move validation to the boundary (`filesclient`).
4. Silent nil-response skip in `readAllContent` — masks impossible state.
5. `grpc.DialContext` is deprecated in grpc v1.79.2 — use `grpc.NewClient`.
6. Missing unit tests for `contenttype.Classify`.

**Minor:**
- Unnecessary nil guards in `parseReadFileInput` for SDK-guaranteed non-nil params.
- `buildContent` uses `default` instead of explicit case for `KindResource` — masks future invariant violations.
- Missing test cases for stream errors, `context.DeadlineExceeded`, and content RPC failures.
- Unreachable `required` guard in Helm `_helpers.tpl`.

**Comment by casey-brooks** [comment]
## Summary
- enforced MaxFileSize invariant in mcpserver, moved metadata validation to filesclient, removed nil fallbacks
- switched gRPC dial to grpc.NewClient and aligned Helm gateway default
- added content-type classifier tests and expanded server error-path coverage

## Test & Lint Summary
- buf generate buf.build/agynio/api --include-imports --path agynio/api/files/v1 --path agynio/api/gateway/v1/files.proto
- go vet ./...
- go test ./...

Tests: passed 12, failed 0, skipped 0
Lint: go vet ./... (no issues)

**Review by noa-lucent — APPROVED** [review]
All 10 review comments verified in actual code and resolved:

1. ✅ Duplicate `defaultMaxFileSize` removed from `server.go` — single source of truth in `config`.
2. ✅ `New()` panics on invalid `MaxFileSize` instead of silently falling back.
3. ✅ Nil metadata check moved to boundary layer (`filesclient.GetFileMetadata`) — internal code no longer has defensive guard.
4. ✅ `parseReadFileInput` simplified to `len(req.Params.Arguments) == 0`.
5. ✅ Silent nil-response skip removed from `readAllContent`.
6. ✅ `buildContent` uses explicit `case contenttype.KindResource` with `default: panic(...)`.
7. ✅ `grpc.DialContext` → `grpc.NewClient`, `ctx` parameter removed from `Dial`, call site updated.
8. ✅ Nil validation for `resp.GetFile()` added in boundary `filesclient.GetFileMetadata`.
9. ✅ Comprehensive `contenttype_test.go` added (16 cases including case-insensitivity, params, edge cases).
10. ✅ Three new test cases: stream error, `context.DeadlineExceeded`, content RPC error. `fakeStream` upgraded with `errIndex` for mid-stream error injection.
11. ✅ Unreachable `required` removed from Helm `_helpers.tpl`.

Clean implementation, no new issues introduced. LGTM.
